### PR TITLE
Dislocation simplification & generalisation

### PR DESCRIPTION
**Pull request by thomas-rocke**
Main Features:
- Simplification of dislocation class interfaces, leveraging dimensionless properties
- Removal of near-identical __init__ methods, leading to improved maintainability
- Generalisation of interface to support arbitrary cubic structures

### CODE REVIEW

**Comment by pgrigorev** [comment]
Thank you for your work, I like the idea of simplification and generalisation! It looks like there is plenty of changes. I will have a closer look this week.

File: matscipy/dislocation.py
Comment (jameskermode) [inline matscipy/dislocation.py:2223]: rename to `_validate_cubic_cell()` and move out of dislocation.py? Perhaps create a new top-level `matscipy.utils` module and import from there to here.

File: matscipy/dislocation.py
Comment (jameskermode) [inline matscipy/dislocation.py:2552]: This could be an abstract base class to prevent it being accidentally instatiated (same for `CubicCrystalDislocation`). I think this is simply `from abc import ABC` then subclass `ABC`.

File: matscipy/dislocation.py
Comment (jameskermode) [inline matscipy/dislocation.py:2588]: check why some attributes come from left and some from right - do we need to check they are consistent?

File: matscipy/dislocation.py
Comment (jameskermode) [inline matscipy/dislocation.py:2706]: pep8: add spaces around `=` operator

File: matscipy/dislocation.py
Comment (jameskermode) [inline matscipy/dislocation.py:2849]: to be moved out of this module

Comment (jameskermode) [reply matscipy/dislocation.py:2849]: `pbc` is initialised to a mutable object. Should be set to `None` and then assigned within function.

File: matscipy/dislocation.py
Comment (jameskermode) [inline matscipy/dislocation.py:2888]: also check array size=0 (i.e. it's a scalar). Check this but I think `np.array(alat).size == 0` when alat is an int or float scalar.

File: tests/test_dislocation.py
Comment (jameskermode) [inline tests/test_dislocation.py:33]: This is a bit odd. Bulk looks like a class name.

File: tests/test_dislocation.py
Comment (jameskermode) [inline tests/test_dislocation.py:422]: This looks like a constructor call. Change to `ase_bulk` or `build_bulk`?

File: tests/test_dislocation.py
Comment (jameskermode) [inline tests/test_dislocation.py:745]: rename to match change in name and location of `_validate_disloc_inputs()`

**Comment by jameskermode** [comment]
Looks good to me with the small comments noted in my review.

@pgrigorev please take a look as well to make sure none of this will impact negatively on your use of this functionality.

After that I'll be happy to merge.

File: matscipy/dislocation.py
Comment (thomas-rocke) [reply matscipy/dislocation.py:2588]: This was reproducing the old init, see the following code (currently line 2755 in dislocations.py):

super().__init__(left_dislocation.unit_cell,
                         left_dislocation.alat,
                         left_dislocation.C11,
                         left_dislocation.C12,
                         left_dislocation.C44,
                         left_dislocation.axes,
                         burgers,
                         unit_cell_core_position=left_dislocation.unit_cell_core_position,
                         parity=left_dislocation.parity,
                         glide_distance=right_dislocation.glide_distance,
                         n_planes=left_dislocation.n_planes,
                         self_consistent=left_dislocation.self_consistent)

Should be safe to change to left disloc glide distance, as there is a check that both are the same.

File: matscipy/dislocation.py
Comment (thomas-rocke) [reply matscipy/dislocation.py]: Changed pbc=True to be the default, which should retain the default behaviour without requiring extra variables.

File: matscipy/dislocation.py
Comment (thomas-rocke) [reply matscipy/dislocation.py]: Adding metaclass=ABCMeta was the solution here, as making CubicCrystalDissociatedDislocation directly subclass ABC created issues in determining the method resolution order (MRO; see https://stackoverflow.com/questions/2010692/what-does-mro-do)

**Comment by thomas-rocke** [comment]
Waiting on #188 to be merged, in order to resolve plasticity documentation between that PR and this one.

File: tests/test_utils.py
Comment (pgrigorev) [inline tests/test_utils.py:80]: What is this print  for?

File: matscipy/dislocation.py
Comment (pgrigorev) [inline matscipy/dislocation.py:2242]: Two spaces

**Comment by pgrigorev** [comment]
Thanks @thomas-rocke  and @jameskermode I had a closer look and it looks great to me. I will update the documentation today and after it can be merged. The documentation page about building cylindrical configurations becomes a bit too long and heavy, especially I after I added FCC and Diamond examples. Thus I think it is better  to move the section **Beyond Single-species Systems** to a separate notebook (and thus separate page in the docs). I think it should be picked up by the table of content in index automatically. We can add referenced between pages as well to make it easier to read.

**Comment by pgrigorev** [comment]
I have just updated #188. As I said before, I think it is best if you put **Beyond Single-species Systems** to a separate notebook in `docs/applications`. Then you just add a line with the name of you notebook after `cylinder_configurations.ipynb` in [docs/applications/plasticity.rst](https://github.com/libAtoms/matscipy/blob/c0bee217596e2ef7e610ea16b3e8a3a7b3f1bce1/docs/applications/plasticity.rst?plain=1#L10) and it should be included in table of contents. 

Once the documentation is synchronised I am happy if this is merged. Great work!

**Comment by thomas-rocke** [comment]
I've merged the WIP changes from #188 in order to resolve the merge conflicts. Should be ready to merge as soon as #188 is approved and merged, assuming no significant changes. However, there could be some value in doing a quick review of my documentation, @jameskermode ? I've done some small bits in `cylinder_configurations.ipynb` documenting some of the new class attributes, and a live list of all of the available dislocation systems. The main part of the docs changes are in `multispecies_dislocations.ipynb`, following the suggestion from @pgrigorev.

I'm also happy to write a little bit for the paper if this work would be a valuable inclusion, but I also don't want to delay the publication at all.

**Comment by jameskermode** [comment]
#188  has now been merged - please update on top of that. If you also merge #198 into this branch then you should get a full set of green ticks from the CI.

**Comment by jameskermode** [comment]
On the JOSS paper, a small textual change to the relevant paragraph (i.e. adding a sentence saying that matscipy can generate dislocations in multi-component systems) won't slow down the review process significantly, so please go ahead and add a possible change to `paper/paper.md` to this PR and I will review it.

File: docs/applications/visualisation.py
Comment (pgrigorev) [inline docs/applications/visualisation.py:133]: I moved this function to `show_dislocation()` that automatically colours the atoms to according to CNA analysis with ovito. Did you try to use it? This also allows to show dislocations as arrows. If it works for your systems it would be nice to use that to be consistent.

**Comment by pgrigorev** [comment]
@thomas-rocke thanks a lot! did you try to build the documentation locally to see how it renders? If not I can do it myself once everything is finished. I have just discovered that there was an error in ovito imports so, the documentation did not render correctly in #188 . It shows a warning in documentation build that I did not see.  I will try to fix it asap. It might be just a question of downgrading the version of ovito.

**Comment by thomas-rocke** [comment]
> @thomas-rocke thanks a lot! did you try to build the documentation locally to see how it renders? If not I can do it myself once everything is finished. I have just discovered that there was an error in ovito imports so, the documentation did not render correctly in #188 . It shows a warning in documentation build that I did not see. I will try to fix it asap. It might be just a question of downgrading the version of ovito.

I haven't tried building the docs locally yet. I presume it's as simple as running the sphinx command that is in the documentation CI? I will also check what your new function does to my systems, but I'm otherwise fairly happy with what the normal `interactive_view` shows. Thanks for all the work in reviewing this - I think the PR is a lot stronger because of it!

**Comment by pgrigorev** [comment]
> > @thomas-rocke thanks a lot! did you try to build the documentation locally to see how it renders? If not I can do it myself once everything is finished. I have just discovered that there was an error in ovito imports so, the documentation did not render correctly in #188 . It shows a warning in documentation build that I did not see. I will try to fix it asap. It might be just a question of downgrading the version of ovito.
> 
> I haven't tried building the docs locally yet. I presume it's as simple as running the sphinx command that is in the documentation CI? I will also check what your new function does to my systems, but I'm otherwise fairly happy with what the normal `interactive_view` shows. Thanks for all the work in reviewing this - I think the PR is a lot stronger because of it!

You are welcome! Yes it should as simple as running sphinx command, but if I remember correctly installing sphinx can be annoying. The new function basically will make an interactive view similar to the picture you posted to slack few weeks back. I think it is worth trying. It makes these visualisations much more informative.

**Comment by thomas-rocke** [comment]
Hi @pgrigorev , I tried to use your `show_dislocation` function, and although I think it's extremely useful for the single-species plots, the colouring of e.g. the diamond lattice region makes the chemical complexity less visible, and therefore makes the plots harder to read. Ideally, I'd like to keep using `interactive_view`, although it is not showing up when I build the docs locally - is this the issue you were talking about?

**Comment by pgrigorev** [comment]
> Hi @pgrigorev , I tried to use your `show_dislocation` function, and although I think it's extremely useful for the single-species plots, the colouring of e.g. the diamond lattice region makes the chemical complexity less visible, and therefore makes the plots harder to read. Ideally, I'd like to keep using `interactive_view`, although it is not showing up when I build the docs locally - is this the issue you were talking about?

The issue I was mentioning was related to dependencies in github actions and was solved in #199, you can see correctly rendered notebook [here](https://libatoms.github.io/matscipy/applications/cylinder_configurations.html#screw-dislocation) now. I think to fix your problem you should add `import nglview` in one of the first cells of th notebook same way it is is done in the `cylinder_configurations.ipynb`. 

I see your point. I would suggest a compromise. We can just add a flag to `show_dislocation` to add CNA coloring. In this way you can keep coloring of the atoms according to chemical species, but the tooltip will show information about identified structure and you will be able to add dislocations lines. I can do it for you. `show_dislocation`  is really a better version of `interactive_view`. It does not make sense to keep both. 

I also noticed that in your branch the `cylinder_configurations.ipynb` has some cells that I deleted in the final version (cells with Vitek plots for examples). Merging does not very well with ipython notebooks. I am afraid you will have to add your cells on top of the final version of the notebook to be sure nothing is lost or broken.

**Comment by thomas-rocke** [comment]
I'd be happy to use `show_dislocation` without the CNA colouring, so this sounds like a good compromise. Thanks for the heads up about the deleted cells as well - I'll work on trying to unpick those merge errors today.

**Comment by thomas-rocke** [comment]
Changes made to include new features of `show_dislocation`, resulting in some really nice plots! I think I made a Git mistake in doing `git checkout --` rather than a `git merge --` to fastforward visualisation.py to include changed from @pgrigorev (I really did not want to have to deal with another merge conflict on the jupyter notebook, the last one took ages to fix!). Should be ready for review now, @jameskermode

File: docs/applications/visualisation.py
Comment (jameskermode) [inline docs/applications/visualisation.py:2]: duplicate import

File: matscipy/dislocation.py
Comment (jameskermode) [inline matscipy/dislocation.py:2162]: Remove `__dict__` from `__slots__`

**Review by jameskermode — APPROVED** [review]
Happy to merge subject to those two minor comments and docs changes

**Comment by pgrigorev** [comment]
The final version of rendered documentation looks fantastic! I like zoomed in visualisations of alpha and better dislocations. Great work, @thomas-rocke. @jameskermode, did we add something to the paper mentioning new functionalities. I added [a comment](https://github.com/openjournals/joss-reviews/issues/5668#issuecomment-1799183999) in the review issue to tell them we updated the docs. We could mention the new feature there as well now.

**Comment by jameskermode** [comment]
Thanks both! Yes, Tom added a remark to the paper too.

**Pull request merged**
